### ISSUE TITLE
Should Chapel support `===`/`!==` operators for checking identity equivalence?

### ISSUE THREAD

**Issue by bradcray** [chapel-lang/chapel#16853]
Swift and Python support distinct `==`/`!=` and `===`/`!==` operators to distinguish between comparing the logical values of classes and their identities.

In Chapel, we have considered disabling `==`/`!=` overloads on classes (see https://github.com/chapel-lang/chapel/issues/13048) in order to avoid breaking the fundamental ability to determine whether one class variable points to the same object as another or contains `nil`.  This issue proposes an alternative, which would be to support both flavors of operators for classes in Chapel where the implementations of `==`/`!=` default to `===`/`!==` for each class if the user hasn't supplied their own overload of `==`/`!=`.  `===`/`!==` would not be permitted to be overloaded.

Advantages:
+ would give users coming from Python the ability to defined `==`/`!=` on their classes as they are accustomed to
+ would make classes more symmetrical in terms of permitting both ordered (e.g., `<`) and unordered (e.g., `==`) comparisons
+ is backwards compatible for user code

Disadvantages:
- it's a change to the language
- internal/standard/package modules would need to be audited to find cases that are expecting `==`/`!=` on classes to be about identity to use `===`/`!==` instead
- could cause confusion to users coming from other languages (like Javascript?) where `===` is supported for values as well (and means "applying no coercions," as I understand it?)

**Comment by mppf** [comment]
What would `===` do on integers or records? (Error? Check that the variable is the same? Or just invoke `==` for those types?)

**Comment by bradcray** [comment]
> What would === do on integers or records? 

I don't have a particularly strong opinion there at present and intentionally wrote this issue to focus strictly on the class case for that reason.  I think there are a number of options available:

* it could just always call `==` for values
* it could compare identity, which might be particularly useful when `ref` arguments are involved
* it could always invoke what today is our compiler-generated/built-in `==` operator, even if a user has overloaded the `===` operator.  In this approach, the compiler would essentially only build `===` operators for types and `==` would call it if the user hadn't defined a `==` operator
* it could be a same-type / no-coercions-permitted comparison, which would be similar thematically to Javascript if I understood my quick read of it properly

**Comment by bradcray** [comment]
Interestingly, if we took option 1 or 3 above, I believe this approach would also fix the problem I had with @e-kayrakli's "`==` overload on classes" case in PR #16824 because the compiler-generated comparison for records would use `===` on fields and therefore my `ChapelBase.chpl` helper routines wouldn't need to have visibility into his source code to resolve properly.  That said, I also believe that case could be addressed by defining comparison operators as methods (and disabling user-defined `==` and `!=` definitely would).

**Comment by mppf** [comment]
I think that having it compare (ref) identity is interesting and what I would lean towards; however that runs the risk of causing confusion with references to a class instance (now are we checking that the references point to the same pointer or that the instances are the same? almost certainly the 2nd but is that confusing? seems no worse than `myC.locale` checking the location of the class fields rather than the location of the variable pointing to the instance)

**Comment by bradcray** [comment]
Interestingly, if we had `===` on values be more about value than identity, it could become the way to check whether two arrays are the same (by some definition of "same") as a single `bool` rather than relying on the somewhat clunky `equals()` method (particularly since classes were the other main case where we've talking about needing/applying `equals()` which becomes less important if the language has two `==`-style operators.  (Issue #11490).

Another potential use of `===` on values is that I believe C++ supports `==` overloads on enums.  If we wanted to do the same, the user could define their own `==` overloads and `===` would check whether or not the symbols were literally the same one (rather than simply equivalent ones).

**Comment by bradcray** [comment]
If we did go with `===` meaning "default / non-overridable comparison for matching value types" (which is where I think I'm currently leaning), we could potentially add support later for comparing the identity/location of values through other (probably yuckier, by design) approaches, such as having a way to say `identity(var1) == identity(var2)` or `sameIdentity(var1, var2)`.  But I'd wait for someone to request this, as I don't think anyone has in the past.

If we went that route, I also think I'd talk about `===` on classes as being less about identity and more about a primitive, rather than semantic, comparison operation (similar to what `==` is today for classes).  I.e., it compares the "raw values" of things, and the "raw value" of a class is its pointer (so therefore its identity).  That would avoid getting into questions like "Why doesn't it check whether the two integer expressions are really the same variable / location?"

**Comment by mppf** [comment]
I guess in that setting then `===` on arrays would indicate if they referred to the same elements (or something along those lines?) which amounts to "pointer identify" AFAIK for arrays?

**Comment by bradcray** [comment]
I was imagining that the "raw value" of an array would be the "raw values" of all of its elements in order to support `A === B` returning `true` for two arrays whose values were `[1.1, 2.2, 3.3]` (and to keep away from identity).  (BTW, I don't particularly like this term "raw value", I was just trying to come up with some differentiator from "logical / semantic value" which is what I would imagine a user overload of `==` on a class or enum might be).

**Comment by mppf** [comment]
I'm not so sure it makes sense to treat `===` on arrays in that way. If `===` on arrays does anything with the elements themselves, don't we have the same motivation to have `===` return an array of bools that motivated `==` doing so? Consider e.g. an array of classes that define `==`. Wouldn't one want to be able to call `===` in a promoted manner (or have something behaving similarly) in the event the test is not meant to run `==`?

In contrast, if `===` on arrays means something about, say, the array buffer pointers being the same, it doesn't do anything for the elements so one would not expect it to promote.

Put another way, I see a big difference between `===` applying to arrays themselves vs. applying to the elements. Having `===` check all of the elements but return `true` seems to me to mix the two a bit. I'd rather have `<==>` do the whole-array thing if we wanted an operator for it. (But I suppose the same arguments can be made there...)

**Comment by bradcray** [comment]
Yeah, I came to a similar conclusion this weekend, though via a slightly different route and planned to capture my thinking this morning.  Here it is:

Assuming we continue to want a compiler-generated `==` operator for records that returns `true` or `false`, for an array field, we have four main options to generate the `bool` for that field:

1. have `==` return a `bool`
2. have `===` return a `bool`
3. have the operator apply a reduction to the `==` expression that returns an array of bools
4. have the operator apply a reduction to the `===` expression that returns an array of bools

Now, let's say the array field is an array of classes and the user has defined an `==` operator for that class type.  Presumably, we'd want any compiler-defined `==` operator on the record to use the user's definition of `==` on the class rather than the built-in "same object" / `===` comparison.

This suggests to me that we can't use approaches 2 or 4 above because I'm assuming that `===` is more primitive than `==` and would only be defined in terms of other `===` operators and never a user-defined `==` operator.

I'm disinclined to take approach 1 for reasons discussed previously (e.g., on issue #11490), which leaves option 3.

This also made me realize that if we support `===` on values, we probably want to have `===` on records be symmetrical to `==` by invoking `===` on each field, which suggests taking approach 4 for the compiler-generated `===` operator on a record.

I'll be curious if you disagree with any of this / it seems off-base to you.

**Comment by mppf** [comment]
> I'll be curious if you disagree with any of this / it seems off-base to you.

Makes sense to me! We can also consider things like `.equals()` or `<==>` as alternatives to doing the reduction in the generated record `==`, but for now I think we should assume we are doing the reduction in the generated `==`.
